Atru/import block ref (#650)
hashicorp/web-unified-docs: content/terraform/v1.12.x/docs/language/block/import.mdx

  Learn how to configure the `import` block so that you can start managing existing resources with Terraform.
---

# `import` block reference

The `import` block instructs Terraform to import existing infrastructure resources into Terraform. Refer to [Import existing resources](/terraform/language/import) for more information.

## Configuration model

An `import` block supports the following configuration:

- [`import`](#import) &nbsp block
  - [`to`](#to) &nbsp address | required
  - [`id`](#id) &nbsp string  
<!--   - [`identity`](#identity) &nbsp map | mutually exclusive with `id` -->
  - [`for_each`](#for_each) &nbsp map or set of strings
  - [`provider`](#provider) &nbsp reference

## Complete configuration

The following `module` block includes all built-in arguments:

```hcl
import {
  to = resource.type.address
  id = "cloud-provider-id"	  
  for_each = {          	  # `for_each` accepts a map or a set of strings 
    <KEY> = <VALUE>
  }
  for_each = [       	  # `for_each` accepts a map or a set of strings 
    "<VALUE>", 
    "<VALUE>"
  ]
  provider = provider-name.alias
}
```

## Specification

An `import` block supports the following configuration.

### `import`

> The `import` block imports existing infrastructure resources into Terraform. You can add an `import` block to any Terraform configuration file, but many organizations either create an `imports.tf` file or place each `import` block beside the destination `resource` block. Refer to [Import existing resources](/terraform/language/import) for more information.

Author: ```suggestion
The `import` block imports existing infrastructure resources into Terraform. You can add an `import` block to any Terraform configuration file, but we recommend either creating an `imports.tf` file for all `import` configurations or placing each `import` block beside the destination `resource` block. Refer to [Import existing resources](/terraform/language/import) for more information.
```

<Tabs>

<Tab heading="List of values">

```hcl
import {
  for_each = [ "<VALUE>" ] 
  # . . .
}
```

</Tab>

<Tab heading="Map of key-value pairs">

```hcl
import {
  for_each = {
    "<KEY>" = "<VALUE>"
  }
}
```

</Tab>

</Tabs>

The `for_each` meta-argument accepts a map or a set of strings and imports an instance for each item in that map or set and produces an `each` object. You can use the values stored in `each.key` and `each.value` for the [`id` argument expression](#id). You can also use them in index expressions in the [`to` argument](#to). This allows Terraform to import multiple instances of a resource configuration.

The inputs for the `for_each` argument must be known for the import plan operation to succeed. You cannot use a `for_each` argument when running `terraform plan` with the `-generate-config-out` option. Refer to [Generate configuration](/terraform/language/import/generating-configuration) for more information.

The `for_each` argument is a meta-argument, which is built into Terraform and controls the way that Terraform creates resources. Refer to [Meta-arguments](/terraform/language/meta-arguments) for more information.

#### Summary

> - Data type: Map or set of strings.

Author: ```suggestion
- Data type: Map or set of strings
```

- Default: None. 
- Example: [Import multiple resources](#import-multiple-resources) 

### `provider`

The `provider` meta-argument instructs Terraform to import resources according to the specified provider configuration.

```hcl
import {
  # . . .
  provider = "<provider>.<alias>"
}
```

By default, Terraform automatically selects a provider based on the resource type, but you can configure multiple instances of the same provider with aliases and use a non-default configuration to import specific resources. 

Use the `<PROVIDER>.<ALIAS>` syntax to reference a provider configuration in the `provider` argument. 

The `provider` argument is a meta-argument, which is built into Terraform and controls the way that Terraform creates resources. Refer to [Meta-arguments](/terraform/language/meta-arguments) for more information.

#### Summary

- Data type: String
- Default: None. 
- Example: [Select an alternate provider configuration](#select-an-alternate-provider-configuration) 

## Examples

The following examples show how to write configuration for common use cases.

### Import a single resource

> In the following example, Terraform imports an AWS instance with the ID `"i-abcd1234"` into the
> `aws_instance.example` resource in the root module:

Author: ```suggestion
In the following example, Terraform imports an S3 bucket to
`aws_s3_bucket.this`:
```

```hcl

> import {
>   to = aws_instance.example
>   id = "i-abcd1234"
> }
> 
> resource "aws_instance" "example" {
>   name = "hashi"

Author: ```suggestion
import {
    to = aws_s3_bucket.this
    id = "example-bucket"
}

resource "aws_s3_bucket" "this" {
}
```
